Develop

Valery223 opened the pull request:
#2

Commit:
Создал слой tranport для book и тест к нему

Commit:
Создал get и  post для `book`  в 3 слоя(handler -> service -> repo).

Valery223 commented:
> ## Pull Request Overview
> 
> This pull request introduces the "Develop" branch changes to implement a basic REST API for managing books, including storage initialization with SQLite, routing configuration using Gin, and layered business logic with service, repository, and transport layers. Key changes include:
> 
>     * Adding SQLite initialization and storage functions.
> 
>     * Implementing routing and HTTP handlers for GET and POST book requests.
> 
>     * Introducing tests for service, transport, and repository layers.
> 
> 
> ### Reviewed Changes
> 
> Copilot reviewed 14 out of 15 changed files in this pull request and generated 2 comments.
> Show a summary per file
> 	
> 	
> 	
> 	
> 	
> 	
> 	
> 	
> 	
> 	
> Files not reviewed (1)
> 
> 
> Comments suppressed due to low confidence (1)

VelvetApparatus commented on cmd/rest-api-main/main.go:
```diff
@@ -13,24 +17,21 @@ func main() {
+	router := router.NewRouter(transport.NewBookHandler(cvs))
 
-	// TODO: init router:gin
+	router.Run(":8080")
```
можно тоже в энвы убрать

Valery223 replied:
исправил

VelvetApparatus commented on internal/book/repository/entity.go:
почему у тебя сущность находится в пакете репозитория? Чем аргументировал ?

Valery223 replied:
каждому слою по структуре?)

VelvetApparatus replied:
ну, сущность это скорее структура отделенная от каких либо слоев. Твой центральная сущность, набор сущностей ядро. А в слоях обозвал бы их моделями

Valery223 replied:
изменил название

VelvetApparatus commented on internal/book/repository/storage.go:
```diff
@@ -0,0 +1,48 @@
+	db *sql.DB
+}
+
+func NewBookStorage(db *sql.DB) *BookStorage {
```
А где интерфейс репо ?

Valery223 replied:
вроде порешали, что использую там , где использую. Если что, можешь дописать

VelvetApparatus replied:
ну, почему ты не интерфейс возвращаешь ?

Valery223 replied:
Будет тогда циклические зависимости

VelvetApparatus commented on internal/book/service/dto.go:
```diff
@@ -0,0 +1,20 @@
+package service
+
+type BookInputDTO struct {
+	ID int
+}
+
+type BookOutputDTO struct {
+	ID     int
+	Name   string
+	Author string
+}
+
+type CreateBookInputDTO struct {
+	Name   string
+	Author string
+}
+
+type CreateBookOutputDTO struct {
+	ID int
+}
```
Input/Output ? Я так понимаю ты так назвал Request/Response ?

Valery223 replied:
Да

VelvetApparatus replied:
а что не request/response ?

Valery223 replied:
я думал в каждом слое есть его входящие и выходящие структуры, которые он поддерживает и остальные должны учитывать это. Типо каждый слой имеет inputDTO и outputDTO. Нормально я не нашел что за DTO

Valery223 replied:
исправил

VelvetApparatus replied:
DTO - Data Transfer Object

VelvetApparatus commented on internal/book/repository/storage.go:
```diff
@@ -0,0 +1,48 @@
+	var bE BookEntity
+	err = stmt.QueryRow(id).Scan(&bE.ID, &bE.Name, &bE.Author)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return BookEntity{}, err
+		}
```
а не стоит ли выдавать кастовую ошибку типо `ErrNotFound`?

Valery223 replied:
вот насчёт ошибок я не знаю как делать. На каждом слое делать кастомные ошибки?

VelvetApparatus replied:
да, в хендлерах можно не делать . А так пусть везде будут

Valery223 replied:
Сделал общие ошибки, пока только для 404

VelvetApparatus commented on internal/book/service/service.go:
```diff
@@ -0,0 +1,45 @@
+	}, nil
+}
+
+func (bs *BookService) PostBook(book CreateBookInputDTO) (CreateBookOutputDTO, error) {
```
`PostBook` ? Мб `CreateBook`. POST это же тупо про REST. У тебя твой service не должен зависеть от транспорта

Valery223 replied:
исправил

VelvetApparatus commented on internal/book/service/service_test.go:
```diff
@@ -0,0 +1,53 @@
+
+type mockStrorage struct{}
+
+func (m *mockStrorage) GetBookById(id int) (repository.BookEntity, error) {
+	if id == 9 {
+		return repository.BookEntity{}, ErrNotFound
+	}
+	return repository.BookEntity{ID: id, Name: "Test Book", Author: "Test Author"}, nil
+}
```
круто, что сразу пишешь с тестами. Это большой плюс. Но большие репо иногда сложно мокать. Посмотри на https://github.com/vektra/mockery . Это код-ген либа, которая за тебя Моки писать будет

Valery223 replied:
Класс, потом посмотрю-переделаю

VelvetApparatus commented on internal/book/service/service.go:
```diff
@@ -0,0 +1,45 @@
+	"learn/rest-api/internal/book/repository"
+)
+
+type Storage interface {
+	GetBookById(id int) (repository.BookEntity, error)
+	SaveBook(repository.BookEntity) (int, error)
+}
```
Storage в service ?

Valery223 replied:
я слышал, что в go нужно интерфейсы писать в местах использования, вот и написал) Не праввильно?

VelvetApparatus replied:
Ну я например пишу  интерфейс рядом с реализациями. А не по месту использования. Тут уже как хочешь.

VelvetApparatus commented on internal/book/transport/handler.go:
```diff
@@ -0,0 +1,58 @@
+		return
+	}
+	input := service.BookInputDTO{ID: query.ID}
+	book, err := h.service.GetBook(input)
+	if err != nil {
+		c.JSON(404, gin.H{"error": err.Error()})
+		return
+	}
```
Ты же не всегда будешь возвращать ошибку, только если книга не найдена. У тебя может слететь коннект к бд или просто какая-то ошибка в запросе. Тогда ты должен возвращать 5хх коды ошибок

Valery223 replied:
исправил

VelvetApparatus reviewed (commented):
gut

Valery223 reviewed (commented):
.

Commit:
Добавил кастомные ошибки, исправил возвращаемые
ошибки(404 и 500). Исправил название структуру
на более подходящие. И некоторые другие доработки

Valery223 reviewed (commented):
.

Valery223 commented on internal/book/service/service_test.go:
```diff
@@ -0,0 +1,53 @@
+	"testing"
+)
+
+type mockStrorage struct{}
```
исправил)

VelvetApparatus commented on internal/storage/sqlite.go:
```diff
@@ -0,0 +1,29 @@
+)
+
+func InitSQLiteDB(dbPath string) (*sql.DB, error) {
+	db, err := sql.Open("sqlite3", dbPath)
```
почему решил взять sqlite ?

Valery223 replied:
потренероваться, и пока разницы не знаю, надо посмотреть

Commit:
Merge branch 'main' into develop
